plugins/crypto/mbedtls/certificategroup.c

```diff
@@ -132,6 +134,108 @@ MemoryCertStore_getRejectedList(UA_CertificateGroup *certGroup, UA_ByteString **
+
+    mbedtls_x509_dn_gets(certSubject, sizeof(certSubject), &cert->subject);
+    mbedtls_x509_dn_gets(crlIssuer, sizeof(crlIssuer), &crl->issuer);
+
+    if(strncmp(certSubject, crlIssuer, MBEDTLS_X509_MAX_DN_NAME_SIZE) == 0) {
+        retval = UA_STATUSCODE_GOOD;
+    } else {
+        retval = UA_STATUSCODE_BADNOMATCH;
+    }
+
+    return retval;
+}
+
+static UA_StatusCode
+mbedtlsFindCrls(UA_CertificateGroup *certGroup, const UA_ByteString *certificate,
+             const UA_ByteString *crlList, const size_t crlListSize,
+             UA_ByteString **crls, size_t *crlsSize) {
+    mbedtls_x509_crt cert;
+    mbedtls_x509_crt_init(&cert);
+    UA_StatusCode retval = UA_mbedTLS_LoadCertificate(certificate, &cert);
+    if(retval != UA_STATUSCODE_GOOD) {
+        UA_LOG_WARNING(certGroup->logging, UA_LOGCATEGORY_SERVER,
+            "An error occurred while parsing the certificate.");
+        return retval;
+    }
+    UA_Boolean foundMatch = false;
+    for(size_t i = 0; i < crlListSize; i++) {
+        mbedtls_x509_crl crl;
+        mbedtls_x509_crl_init(&crl);
+        retval = UA_mbedTLS_LoadCrl(&crlList[i], &crl);
+        if(retval != UA_STATUSCODE_GOOD) {
+            UA_LOG_WARNING(certGroup->logging, UA_LOGCATEGORY_SERVER,
+                "An error occurred while parsing the crl.");
+            mbedtls_x509_crt_free(&cert);
+            return retval;
+        }
+
+        retval = mbedtlsCheckCrlMatch(&cert, &crl);
+        mbedtls_x509_crl_free(&crl);
+
+        if(retval == UA_STATUSCODE_BADNOMATCH) {
+            continue;
+        }
+        /* If it is not a CA certificate, there is no crl list. */
+        if(retval == UA_STATUSCODE_BADCERTIFICATEUSENOTALLOWED) {
+            UA_LOG_WARNING(certGroup->logging, UA_LOGCATEGORY_SERVER,
+                "The certificate is not a CA certificate and therefore does not have a CRL.");
+            mbedtls_x509_crt_free(&cert);
+            return retval;
+        }
+        /* Continue the search, as a certificate may be associated with multiple CRLs. */
+        foundMatch = true;
+        retval = UA_Array_appendCopy((void **)crls, crlsSize, &crlList[i],
+                                 &UA_TYPES[UA_TYPES_BYTESTRING]);
+        if(retval != UA_STATUSCODE_GOOD) {
+            mbedtls_x509_crt_free(&cert);
+            return retval;
+        }
+    }
+    mbedtls_x509_crt_free(&cert);
+
+    if(!foundMatch)
+        return UA_STATUSCODE_BADNOMATCH;
+
+    return UA_STATUSCODE_GOOD;
+}
+
+static UA_StatusCode
+MemoryCertStore_getCertificateCrls(UA_CertificateGroup *certGroup, const UA_ByteString *certificate,
+                                   UA_ByteString **crls, size_t *crlsSize) {
+    /* Check parameter */
+    if(certGroup == NULL || certificate == NULL || crls == NULL) {
+        return UA_STATUSCODE_BADINTERNALERROR;
+    }
+
+    MemoryCertStore *context = (MemoryCertStore *)certGroup->context;
+
+    UA_StatusCode retval = mbedtlsFindCrls(certGroup, certificate,
```

Comment:
Is it possible that - by mistake or by design - a CRL is both in the issuer and the revocation list?
For example when an issuer gets moved to the trustlist.
We should be concatenating the two lists.

But we can also simplify overall.
Please keep a single CRL list for both issuers and trustlist.
The functionality is identical as we map to issuer names anyhow.

Reply:
Based on my understanding of the specification, it's possible for a CA certificate to appear on both the Issuer and Trusted lists. The Trusted list contains certificates that are explicitly trusted, which may include CA certificates. The Issuer list, on the other hand, includes certificates necessary for validation purposes.

For this reason, I believe it’s important to maintain a clear distinction between trusted and issuer CRLs. If a CA certificate is removed from the trusted list, only the CRLs associated with it should be removed from the trusted list, while the Issuer list should remain unaffected.

This separation wouldn’t be feasible if there were only one combined CRL list. Additionally, the TrustListDataType is designed to support this distinction.

Reply:
I have modified the getCertificateCrls function by introducing the isTrusted parameter, allowing you to specify which list to reference. Since a certificate may appear in both lists, this parameter ensures the correct list is selected.

Additionally, a certificate can also have multiple CRLs, and the trustedCrls list and the issuerCrls list may each contain different CRLs for the same certificate.